strawberry/experimental/pydantic2/object_type.py

```diff
@@ -0,0 +1,343 @@
+        field.default_factory is not None or field.default is not None
+    )
+    should_add_optional: bool = not (is_required(field) or default_defined)
+    if should_add_optional:
+        return Optional[replaced_type]
+    else:
+        return replaced_type
+
+
+def _build_dataclass_creation_fields(
+    field_name: str,
+    field: FieldInfo,
+    is_input: bool,
+    existing_fields: Dict[str, StrawberryField],
+    auto_fields_set: Set[str],
+    use_pydantic_alias: bool,
+) -> DataclassCreationFields:
+    field_type = (
+        get_type_for_field(field, is_input)
+        if field_name in auto_fields_set
+        else existing_fields[field_name].type
+    )
+
+    if (
+        field_name in existing_fields
+        and existing_fields[field_name].base_resolver is not None
+    ):
+        # if the user has defined a resolver for this field, always use it
+        strawberry_field = existing_fields[field_name]
+    else:
+        # otherwise we build an appropriate strawberry field that resolves it
+        existing_field = existing_fields.get(field_name)
+        graphql_name = None
+        if existing_field and existing_field.graphql_name:
+            graphql_name = existing_field.graphql_name
+        elif field.alias and use_pydantic_alias:
+            graphql_name = field.alias
+
+        strawberry_field = StrawberryField(
+            python_name=field_name,
+            graphql_name=graphql_name,
+            # always unset because we use default_factory instead
+            default=dataclasses.MISSING,
+            default_factory=get_default_factory_for_field(field),
+            type_annotation=StrawberryAnnotation.from_annotation(field_type),
+            description=field.description,
+            deprecation_reason=(
+                existing_field.deprecation_reason if existing_field else None
+            ),
+            permission_classes=(
+                existing_field.permission_classes if existing_field else []
+            ),
+            directives=existing_field.directives if existing_field else (),
+            metadata=existing_field.metadata if existing_field else {},
+        )
+
+    return DataclassCreationFields(
+        name=field_name,
+        field_type=field_type,
+        field=strawberry_field,
+    )
+
+
+if TYPE_CHECKING:
+    from strawberry.experimental.pydantic2.conversion_types import (
+        PydanticModel,
+        StrawberryTypeFromPydantic,
+    )
+
+
+def type(
+    model: Type[PydanticModel],
+    *,
+    name: Optional[str] = None,
+    is_input: bool = False,
+    is_interface: bool = False,
+    description: Optional[str] = None,
+    directives: Optional[Sequence[object]] = (),
+    all_fields: bool = False,
+    use_pydantic_alias: bool = True,
+) -> Callable[..., Type[StrawberryTypeFromPydantic[PydanticModel]]]:
+    def wrap(cls: Any) -> Type[StrawberryTypeFromPydantic[PydanticModel]]:
+        model_fields: Dict[str, FieldInfo] = model.model_fields
```

Comment:
in pydantic v1 this was ModelFields, but its now FieldInfo in pydantic v2. The `FieldInfo` also does not contain the name of the field itself, so now its a dit.